storage/src/adb/any/variable/mod.rs

```diff
@@ -450,80 +386,53 @@ impl<E: RStorage + Clock + Metrics, K: Array, V: Codec, H: CHasher, T: Translato
     ///
     /// Returns [crate::mmr::Error::LocationOverflow] if `loc` > [crate::mmr::MAX_LOCATION].
     /// Returns [Error::LocationOutOfBounds] if `loc` >= [Self::op_count].
+    /// Returns [Error::UnexpectedData] if the location does not reference an Update operation.
     pub async fn get_from_loc(&self, key: &K, loc: Location) -> Result<Option<V>, Error> {
         if !loc.is_valid() {
             return Err(Error::Mmr(crate::mmr::Error::LocationOverflow(loc)));
         }
-        if loc >= self.op_count() {
-            return Err(Error::LocationOutOfBounds(loc, self.op_count()));
+        let op_count = self.op_count().await?;
+        if loc >= op_count {
+            return Err(Error::LocationOutOfBounds(loc, op_count));
         }
 
-        match self.locations.read(*loc).await {
-            Ok(offset) => {
-                return self.get_from_offset(key, loc, offset).await;
-            }
-            Err(e) => Err(Error::Journal(e)),
+        let op = self.get_op(loc).await?;
+        let Operation::Update(k, v) = op else {
+            return Err(Error::UnexpectedData(loc));
+        };
+        if k != *key {
+            return Ok(None);
         }
+        Ok(Some(v))
     }
 
     /// Get the operation at location `loc` in the log.
     async fn get_op(&self, loc: Location) -> Result<Operation<K, V>, Error> {
-        match self.locations.read(*loc).await {
-            Ok(offset) => {
-                let section = *loc / self.log_items_per_section;
-                self.log.get(section, offset).await.map_err(Error::Journal)
-            }
-            Err(e) => Err(Error::Journal(e)),
-        }
-    }
-
-    /// Get the value of the operation with location `loc` and offset `offset` in the log if it
-    /// matches `key`.
-    ///
-    /// # Errors
-    ///
-    /// Returns [Error::UnexpectedData] if the location does not reference an Update operation.
-    async fn get_from_offset(
-        &self,
-        key: &K,
-        loc: Location,
-        offset: u32,
-    ) -> Result<Option<V>, Error> {
-        let section = *loc / self.log_items_per_section;
-        let Operation::Update(k, v) = self.log.get(section, offset).await? else {
-            return Err(Error::UnexpectedData(loc));
-        };
-
-        if k != *key {
-            Ok(None)
-        } else {
-            Ok(Some(v))
-        }
+        self.log.read(*loc).await.map_err(Error::Journal)
     }
 
     /// Get the number of operations that have been applied to this db, including those that are not
     /// yet committed.
-    pub fn op_count(&self) -> Location {
-        self.log_size
+    pub async fn op_count(&self) -> Result<Location, Error> {
+        self.log
+            .size()
```

Comment:
is this going to be slow? (e.g. involve a file syscall?). if so maybe we should keep the log_size cached?

Reply:
huh, where does the potential Error come from?

Reply:
You're right, this doesn't need to be `async` or `Result`. [Changed](https://github.com/commonwarexyz/monorepo/pull/2036#discussion_r2478877875) the signature here.